src/grpc/iot_config/hpr_route_ets.erl

```diff
@@ -191,62 +217,105 @@ lookup_devaddr_range(DevAddr) ->
+            }),
+            lager:debug(
+                [
+                    {route_id, RouteID},
+                    {devaddr, hpr_utils:int_to_hex_string(DevAddr)},
+                    {session_key, SessionKey},
+                    {max_copies, MaxCopies}
+                ],
+                "inserting SKF"
+            );
+        _Other ->
+            lager:error("failed to insert skf table not found ~p for ~s", [
+                _Other, RouteID
+            ])
+    end,
+    ok.
+
+-spec update_skf(
+    DevAddr :: non_neg_integer(),
+    SessionKey :: binary(),
+    RouteID :: string(),
+    MaxCopies :: non_neg_integer()
+) -> ok.
+update_skf(DevAddr, SessionKey, RouteID, MaxCopies) ->
+    SKF = hpr_skf:new(#{
+        route_id => RouteID,
+        devaddr => DevAddr,
+        session_key => hpr_utils:bin_to_hex_string(SessionKey),
+        max_copies => MaxCopies
+    }),
+    ok = delete_skf(SKF),
+    ok = insert_skf(SKF),
     ok.
 
 -spec delete_skf(SKF :: hpr_skf:skf()) -> ok.
 delete_skf(SKF) ->
-    RouteId = hpr_skf:route_id(SKF),
-    DevAddr = hpr_skf:devaddr(SKF),
-    SessionKey = hpr_skf:session_key(SKF),
-    MaxCopies = hpr_skf:max_copies(SKF),
-    true = ets:delete_object(
-        ?ETS_SKFS, {DevAddr, {hpr_utils:hex_to_bin(SessionKey), RouteId, MaxCopies}}
-    ),
-    lager:debug(
-        [
-            {route_id, RouteId},
-            {devaddr, hpr_utils:int_to_hex_string(DevAddr)},
-            {session_key, SessionKey},
-            {max_copies, MaxCopies}
-        ],
-        "deleting SKF"
-    ),
+    RouteID = hpr_skf:route_id(SKF),
+    case ?MODULE:lookup_route(RouteID) of
+        [{_, SKFETS}] ->
+            DevAddr = hpr_skf:devaddr(SKF),
+            SessionKey = hpr_skf:session_key(SKF),
+            MaxCopies = hpr_skf:max_copies(SKF),
+            %% Here we ignore max_copies
+            MS = [{{{'_', hpr_utils:hex_to_bin(SessionKey)}, {DevAddr, '_'}}, [], [true]}],
+            Deleted = ets:select_delete(SKFETS, MS),
+            lager:debug(
+                [
+                    {route_id, RouteID},
+                    {devaddr, hpr_utils:int_to_hex_string(DevAddr)},
+                    {session_key, SessionKey},
+                    {max_copies, MaxCopies}
+                ],
+                "deleted ~p SKF",
+                [Deleted]
+            );
+        _Other ->
+            lager:warning("failed to delete skf not found ~p for ~s", [
+                _Other, RouteID
+            ])
+    end,
     ok.
 
--spec lookup_skf(DevAddr :: non_neg_integer()) -> [{binary(), string(), non_neg_integer()}].
-lookup_skf(DevAddr) ->
-    case ets:lookup(?ETS_SKFS, DevAddr) of
-        [] -> [];
-        SKFs -> [{Filter, RouteID, MaxCopies} || {_DevAddr, {Filter, RouteID, MaxCopies}} <- SKFs]
-    end.
+-spec lookup_skf(ETS :: ets:table(), DevAddr :: non_neg_integer()) ->
+    [{binary(), integer(), non_neg_integer()}].
```

Comment:
```suggestion
    [{SessionKey :: binary(), Timestamp :: integer(), MaxCopies :: non_neg_integer()}].
```
As a helper because I had a hard time reading the numbers swapped around.